convert2rhel/checks.py

```diff
@@ -16,13 +16,190 @@
+    def check_output(command):
+        return subprocess.Popen(
+            command,
+            stdout=subprocess.PIPE,
+        ).communicate()[0]
+
+
+from convert2rhel.systeminfo import system_info
+from convert2rhel.utils import run_subprocess
 
-from convert2rhel.pkghandler import call_yum_cmd
 
 logger = logging.getLogger(__name__)
 
+KERNEL_REPO_RE = re.compile("^.+:(?P<version>.+).el.+$")
+KERNEL_REPO_VER_SPLIT_RE = re.compile("\W+")
+LINK_KMODS_RH_POLICY = "https://access.redhat.com/third-party-software-support"
+
+
+def _convert_to_int_or_zero(s):
+    try:
+        return int(s)
+    except ValueError:
+        return 0
+
+
+def _get_kmod_comparison_key(path):
+    """Create a comparison key from the kernel module abs path.
+
+    Converts /lib/modules/5.8.0-7642-generic/kernel/lib/a.ko.xz ->
+    kernel/lib/a.ko.xz
+
+    Why:
+        all kernel modules lives in /lib/modules/{some kernel release}
+        if we want to make sure that the kernel package is really presented
+        on RHEL, we need to compare the full path, but because kernel release
+        might be different, we compare the relative paths after kernel release.
+    """
+    return "/".join(path.split("/")[4:])
+
+
+def get_host_kmods():
+    try:
+        # TODO make it work with utils.run_subprocess (now fails)
+        kmod_str = check_output(
+            'find /lib/modules/"$(uname -r)" -name "*.ko*"',
+            shell=True,
+        ).decode()
+        assert kmod_str
+    except (subprocess.CalledProcessError, AssertionError):
+        logger.critical("Can't get list of kernel modules.")
+    else:
+        return set(
+            _get_kmod_comparison_key(path)
+            for path in kmod_str.rstrip("\n").split()
+        )
+
+
+def _repos_version_key(pkg_name):
+    try:
+        rpm_version = KERNEL_REPO_RE.search(pkg_name).group("version")
+    except AttributeError:
+        raise NotImplementedError(
+            "Unexpected package:\n%s\n is a source of kernel modules."
+            % pkg_name
+        )
+    else:
+        return tuple(
+            map(
+                _convert_to_int_or_zero,
+                KERNEL_REPO_VER_SPLIT_RE.split(rpm_version),
+            )
+        )
+
+
+def get_most_recent_unique_kernel_pkgs(pkgs):
+    """Return the most recent versions of all kernel packages.
+
+    When we do scanning of kernel modules provided by kernel packages,
+    it is expensive to check each kernel pkg. Considering the fact,
+    that each new kernel pkg do not deprecate kernel modules we select only
+    the most recent ones.
+
+    :type pkgs: Iterable[str]
+    """
+
+    pkgs_groups = itertools.groupby(
+        pkgs, lambda pkg_name: pkg_name.split(":")[0]
+    )
+    return (
+        max(distinct_kernel_pkgs[1], key=_repos_version_key)
+        for distinct_kernel_pkgs in pkgs_groups
+        if distinct_kernel_pkgs[0].startswith(("kernel", "kmod"))
+    )
+
+
+def get_rhel_supported_kmods(_refresh_system_info=False):
```

Comment:
Could use more inline comments to explain what some pieces of code does

Reply:
yeap